journalbeat/input/input.go

```diff
@@ -0,0 +1,174 @@
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+// Input manages readers and forwards entries from journals.
+type Input struct {
+	readers  []*reader.Reader
+	done     chan struct{}
+	config   Config
+	pipeline beat.Pipeline
+	states   map[string]checkpoint.JournalState
+	id       uuid.UUID
+	logger   *logp.Logger
+}
+
+// New returns a new Inout
+func New(
+	c *common.Config,
+	pipeline beat.Pipeline,
+	done chan struct{},
+	states map[string]checkpoint.JournalState,
+) (*Input, error) {
+	config := DefaultConfig
+	if err := c.Unpack(&config); err != nil {
+		return nil, err
+	}
+
+	id := uuid.NewV4()
+	logger := logp.NewLogger("input").With("id", id)
+
+	var readers []*reader.Reader
+	if len(config.Paths) == 0 {
+		cfg := reader.Config{
+			Path:          reader.LocalSystemJournalID, // used to identify the state in the registry
+			Backoff:       config.Backoff,
+			MaxBackoff:    config.MaxBackoff,
+			BackoffFactor: config.BackoffFactor,
+			Seek:          config.Seek,
+		}
+
+		state := states[reader.LocalSystemJournalID]
+		r, err := reader.NewLocal(cfg, done, state, logger)
+		if err != nil {
+			return nil, fmt.Errorf("error creating reader for local journal: %v", err)
+		}
+		readers = append(readers, r)
+	}
+
+	for _, p := range config.Paths {
+		cfg := reader.Config{
+			Path:          p,
+			Backoff:       config.Backoff,
+			MaxBackoff:    config.MaxBackoff,
+			BackoffFactor: config.BackoffFactor,
+			Seek:          config.Seek,
+		}
+		state := states[p]
+		r, err := reader.New(cfg, done, state, logger)
+		if err != nil {
+			return nil, fmt.Errorf("error creating reader for journal: %v", err)
+		}
+		readers = append(readers, r)
+	}
+
+	logger.Debugf("New input is created for paths %v", config.Paths)
+
+	return &Input{
+		readers:  readers,
+		done:     done,
+		config:   config,
+		pipeline: pipeline,
+		states:   states,
+		id:       id,
+		logger:   logger,
+	}, nil
+}
+
+// Run connects to the output, collects entries from the readers
+// and then publishes the events.
+func (i *Input) Run() {
+	client, err := i.pipeline.ConnectWith(beat.ClientConfig{
+		PublishMode:   beat.GuaranteedSend,
+		EventMetadata: common.EventMetadata{},
+		Meta:          nil,
+		Processor:     nil,
+		ACKCount: func(n int) {
+			i.logger.Infof("journalbeat successfully published %d events", n)
+		},
+	})
+	if err != nil {
+		i.logger.Error("Error connecting to output: %v", err)
+		return
+	}
+	defer client.Close()
+
+	for {
+		select {
+		case <-i.done:
+			return
+		default:
+			i.publishAll(client)
+		}
+	}
+
+}
+
+func (i *Input) publishAll(client beat.Client) {
+	out := make(chan *beat.Event)
+	var wg sync.WaitGroup
+
+	merge := func(in chan *beat.Event) {
+		wg.Add(1)
+
+		go func(c chan *beat.Event) {
+			for v := range c {
+				out <- v
+			}
+			wg.Done()
+		}(in)
+	}
+	go func() {
+		wg.Wait()
+		close(out)
+	}()
+
+	for _, r := range i.readers {
+		c := r.Follow()
+		merge(c)
+	}
+
+	for e := range out {
+		client.Publish(*e)
+	}
+}
```

Comment:
The way the code is structured, `publishAll` should be a blocking call and we should do the check if `is.Done` is called here. 

Maybe something like the following will work.

```golang
func (i *Input) publishAll(client beat.Client) {
	out := make(chan *beat.Event)
	defer close(out)

	var wg sync.WaitGroup
	merge := func(in chan *beat.Event) {
		wg.Add(1)
		go func(in chan *beat.Event, out chan *beat.Event) {
			for {
				select {
				case <-i.Done:
					return
				case out <- c:
				}
			}
		}(in)
	}

	for _, r := range i.readers {
		c := r.Follow()
		merge(c)
	}

	for {
		select {
		case <-is.Done:
			break
		case e <- out:
			client.Publish(*e)
		}
	}

	wg.Wait()
}